projects/wallet-service-reliable-integration/0001-design.md

```diff
@@ -0,0 +1,205 @@
+**Transitions**:
+- If the service invocation completes successfully (`onDone`):
+    - **Target**: The machine transitions to the `CONNECTED.success` sub-state.
+- If there's an error during the service invocation (`onError`):
+    - **Target**: The machine transitions to the `CONNECTED.error` sub-state.
+
+
+#### `CONNECTED.success`
+
+The machine will transition to this state when a event is successfully processed
+
+**Entry Action**: As soon as the machine enters the `success` state, it sends an acknowledgment using the `sendAck` action.
+**Transitions**:
+  - The machine always transitions back to the `CONNECTED.idle` sub-state after performing the entry actions.
+
+
+#### `CONNECTED.error`
+
+The machine will transition to this state when an error is detected when event handling or invoking a `service`.
+
+This is a `final` state, meaning that no further transitions can occur from this state and that the machine will need manual intervention to recover.
+
+**Entry Action**: The machine logs the error using the `logError` action.
+
+
+## Potential challenges
+
+1. **Event Ordering and Consistency**: Given that events are sent in the order they occurred, it's crucial to ensure that the daemon maintains this order when processing events. This is especially important in scenarios like reorgs where the event order can significantly impact the wallet-service's state.
+2. **Event Volume and Scalability**: Depending on the full-node's activity, the number of events generated can be significant and considering that we are processing those events one at a time, we should make sure that we are doing this as efficiently as possible
+3. **Error Handling and Recovery**: There might be situations where the daemon encounters an error processing an event or loses connection to the full-node. We should make sure that we always have the latest successfully synced event id persisted and that we are rolling back changes that were unsuccessful to prevent invalid states
+
+
+# Guide-level explanation
+
+## Actors
+
+> The [Actor model (opens new window)](https://en.wikipedia.org/wiki/Actor_model)is a mathematical model of message-based computation that simplifies how multiple "entities" (or "actors") communicate with each other. Actors communicate by sending messages (events) to each other. An actor's local state is private, unless it wishes to share it with another actor, by sending it as an event.
+https://xstate.js.org/docs/guides/actors.html
+
+
+#### WebSocketActor
+
+The `WebSocketActor` will initiate and maintain a connection with the full-node's WebSocket feed.
+
+We are using actors to have a bi-directional channel to the WebSocket feed from the machine, it will receive events and send those events to the machine, using the same type as received from the feed, so if new events are added in the future, they will be automatically emitted to the machine (and probably handled by the '\*' transition)
+
+
+## Services
```

Comment:
If these methods are well tested, I feel it's ok to move forward with them. But I think there might a simpler way to handle these events. Let me explain.

Suppose events are group by `group_id`, and that you can create a new database transaction for each group, only committing after the last event of a group is processed. In this scenario, you can safely do the following:

1. When a group is started, you run a `START TRANSACTION` on the database.
2. Then, you process each transaction independently, updating the utxo table according to the event information. In other words, if an output is unspent, it goes into the utxo table. If its spent, you delete its related utxo. For clarity, you don't have to do anything related to inputs or parents.
2. You can process events in parallel (unless they are related to the same transaction).
3. After the whole group is processed, you can generate new addresses, unlock timelocked utxos and recalculate balance only once.
4. Finally, you `COMMIT` the changes to the database.

What do you think? Maybe we should consider doing something like this in the future since your current implementation is well tested.

Reply:
@msbrogli I added two future possibilities with the optimizations we discussed in our last call

Reply:
Sounds great. Thanks!